samtranslator/model/sam_resources.py

```diff
@@ -2540,6 +2539,100 @@ def _construct_ddb_datasource_connector_resources(
         )
         return connector.to_cloudformation(**kwargs)
 
+    def _construct_lambda_datasources(
+        self,
+        lambda_datasources: Optional[Dict[str, aws_serverless_graphqlapi.LambdaDataSource]],
+        api_id: Intrinsicable[str],
+        kwargs: Dict[str, Any],
+    ) -> List[Resource]:
+        if not lambda_datasources:
+            return []
+
+        resources: List[Resource] = []
+
+        for relative_id, lambda_datasource in lambda_datasources.items():
+            datasource_logical_id = f"{self.logical_id}Lambda{relative_id}"
+            cfn_datasource = DataSource(
+                logical_id=datasource_logical_id, depends_on=self.depends_on, attributes=self.resource_attributes
+            )
+
+            cfn_datasource.Name = lambda_datasource.Name or relative_id
```

Comment:
Should this be unique or is `relative_id` enough? (i.e. you can have multiple data sources with the same `Name`)

Reply:
I think I remember us discussing this before with DDB datasources. We said that it can feel weird for the customer if the default value (relative_id in this case) to not be exactly what the customer put, so we left it as `relative_id`, even though it may cause unique conflicts.

Reply:
what kind of conflicts does it cause?

Reply:
Does it cause deployment failure? Specifically, I want us to avoid and be aware of bugs like https://github.com/aws/serverless-application-model/issues/3109. These have occurred a few times.

Reply:
DataSourceName has to be unique in that region. So two datasources cannot share the same name.

Reply:
Okay so this will cause deployment issues quite easily? Whether I deploy the same template to different stacks, or have the same `relative_id` (which I assume is the key in a dict? Similar to `Events`) in different resources. If yes, it will be a frustrating experience and we should use something more unique.

Reply:
Offline discussion: Because `DataSourceName` is a unique identifier for datasources, the way we currently default them will cause customer issues like in #3109. We are going to investigate a way to create a unique DataSourceName tied to the stack name such that customers won't have any issue redeploying the same template to a different stack.